### PR TITLE
[tt-train] Python Implementation of TTML LR Schedulers

### PR DESCRIPTION
**Pull request by awliu-TT**
### Summary
Currently, ttml does not expose the learning rate schedulers to Python. Also, there is no implementation of the CosineAnnealingLR scheduler widely used in the pretraining of LLMs. This PR adds Python implementations of the ttml learning rate schedulers and adds an implementation of the CosineAnnealingLR scheduler. Unit tests were added to all Python lr schedulers to test for correctness and working `state_dict` saving (used to restore checkpoints during training).

This PR also changes the C++ Sequential LR Scheduler behavior from saving and restoring only the `state_dict` of the current LR scheduler to saving and restoring the `state_dict` of all LR schedulers. This matches PyTorch behavior, and allows the states of future LR schedulers in the sequence to be restored for resuming training.

### Checklist

- [x] [Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/25772471898) - [main results](https://github.com/tenstorrent/tt-metal/actions/runs/25093195989)
- [x] [Blackhole Post Commit](https://github.com/tenstorrent/tt-metal/actions/runs/25772458288) - [main results](https://github.com/tenstorrent/tt-metal/actions/runs/25091794658)
- [x] [Nightly L2](https://github.com/tenstorrent/tt-metal/actions/runs/25772427854) - [main results](https://github.com/tenstorrent/tt-metal/actions/runs/25718556230/job/75554194315)
- [x] [(Runtime) Sanity Tests](https://github.com/tenstorrent/tt-metal/actions/runs/25772443742)
- [x] New/Existing unit tests for this change

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:awliu/cosine_annealing_lr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:awliu/cosine_annealing_lr)

### CODE REVIEW

File: tt-train/tests/python/test_schedulers.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+  - TestXMatchesPyTorch  — trajectory compared step-by-step against the
+                            equivalent torch.optim.lr_scheduler class.
+  - TestXStateDict        — get_state_dict / set_state_dict round-trip.
+"""
+
+import pytest
+import torch
+import ttml
+from ttml.common.schedulers import CosineAnnealingScheduler, LambdaScheduler, LinearScheduler, StepScheduler
+
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/tests/python/test_schedulers.py:28]: nanobind exposure removed.

File: tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+serialization::StateDict CosineAnnealingScheduler::get_state_dict() const {
+    serialization::StateDict res;
+    res["m_last_step"] = m_last_step;
+    res["m_last_lr"] = m_last_lr;
+    return res;
+}
+
+void CosineAnnealingScheduler::set_state_dict(const serialization::StateDict& dict) {
+    m_last_step = serialization::get_value_type<size_t>(dict, "m_last_step");
+    m_last_lr = serialization::get_value_type<float>(dict, "m_last_lr");
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp:56]: This is actually something that should be done by the optimizer when it restores from its `state_dict`.

File: tt-train/sources/ttml/ttml/common/schedulers.py [hunk 3]
```diff
@@ -53,3 +49,93 @@ def __init__(self, optim):
...
+
+    def get_current_lr(self) -> float:
+        return self._optimizer.get_lr()
+
+    def get_state_dict(self) -> dict:
+        return {"m_last_step": self._last_step, "m_last_lr": self._last_lr}
+
+    def set_state_dict(self, state: dict):
+        self._last_step = state["m_last_step"]
+        self._last_lr = state["m_last_lr"]
```
[the remark this answers is not in the document]
Comment (awliu-TT) [reply tt-train/sources/ttml/ttml/common/schedulers.py:87]: This is actually something that should be done by the optimizer when it restores from its `state_dict`.

**Review by philei-tt — COMMENTED** [review]
Thanks, looks good. 

1. We already have some schedulers (identity and linear) in cpp training loop: `tt-train/sources/examples/nano_gpt/utils.cpp`. Please remove those and update cpp training loop. Make sure loss curve is identical before / after (on good amount of steps).
2. I don't understand why are you reimplementing schedulers in python. Either pybind cpp schedulers or remove them altogether??

**Comment by awliu-TT** [comment]
> Thanks, looks good.
> 
> 1. We already have some schedulers (identity and linear) in cpp training loop: `tt-train/sources/examples/nano_gpt/utils.cpp`. Please remove those and update cpp training loop. Make sure loss curve is identical before / after (on good amount of steps).
> 2. I don't understand why are you reimplementing schedulers in python. Either pybind cpp schedulers or remove them altogether??

Regarding 1., sure I can try that. Regarding 2., I was informed that there was an effort in flight to remove the cpp implementations eventually. Since things like LR schedulers are not very performance-critical, I was given the understanding that we would be switching to pure-python implementations like how it is in PyTorch. I wasn't sure if we still intended to keep the cpp implementations in the meantime, but if we can get rid of them now I'd be happy to remove those too.

**Review by dyurshevichTT — COMMENTED** [review]
Also looking good to me, good job! Also would want to double check the learning curves, but test coverage is good so don't expect anything to fail. Regarding C++ not sure we want to delete schedulers now as without schedulers it doesn't make much sense to farther support C++ training loop. But if we have support of all the cpp staff in python already, then we can agree on abandoning C++.

File: tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
```
Comment (jmalone-tt) [inline tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp:1]: Do we need to add CPP versions of new schedulers, or is it better to just use python going forward?

**Review by jmalone-tt — COMMENTED** [review]
Overall looks good to me, but I think we should discuss whether or not we need to add the cpp implementation

**Comment by awliu-TT** [comment]
@philei-tt @dyurshevichTT @jmalone-tt Actually, are we really sure we want to get rid of the C++ implementations? PyTorch maintains two separate parallel implementations of their LR schedulers. They have a pure Python implementation as well as C++ implementations for their LibTorch C++ distribution: https://docs.pytorch.org/cppdocs/api/optim/schedulers.html. Since we already have these LR schedulers implemented in C++, it might be useful to keep them. I suppose the real question is do we have any intention or plans to continue supporting C++ training frameworks, e.g., for devs who want to integrate training in some C++ environment? If we want to keep the C++ implementation but maintain Python access without two parallel implementations, I could also just revert to exposing the C++ schedulers through nanobind like in [6ca1897](https://github.com/tenstorrent/tt-metal/pull/43469/commits/6ca189790d5afcc6eae34012b99813e6dc883f4a). What do you think?

**Comment by philei-tt** [comment]
It's much easier and faster to add features in python, and we try to keep only perf critical stuff in cpp. We are still keeping cpp training loop up to date till we port and test all features in python.

Honestly, I see zero issues with having cpp scheduler, its simple and I don't see this as something that will be hard to maintain or something that will be limiting adding new things in the future

**Comment by awliu-TT** [comment]
> It's much easier and faster to add features in python, and we try to keep only perf critical stuff in cpp. We are still keeping cpp training loop up to date till we port and test all features in python.
> 
> Honestly, I see zero issues with having cpp scheduler, its simple and I don't see this as something that will be hard to maintain or something that will be limiting adding new things in the future

Sounds good. In that case I'll keep both then. I do see the value in having pure-python implementations. Most people will be using the python infra, and adding features there will be much easier.

File: tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp (hunk 4 above)
Comment (awliu-TT) [reply tt-train/sources/ttml/schedulers/cosine_annealing_scheduler.cpp:1]: Since we are keeping both cpp and python schedulers, I figured I'll just keep the cpp cosine lr scheduler, since I've already implemented and tested it anyway. It's a fairly common scheduler.

File: tt-train/sources/ttml/ttml/common/schedulers.py [hunk 5]
```diff
@@ -53,3 +52,239 @@ def __init__(self, optim):
...
+        self._current_step_in_scheduler += 1
+        self._last_lr = current.get_last_lr()
+
+        if self._current_step_in_scheduler >= current_sched_steps:
+            self._current_scheduler_index += 1
+            self._current_step_in_scheduler = 0
+
+    def get_last_lr(self) -> float:
+        if self._current_scheduler_index < len(self._schedulers):
+            return self._schedulers[self._current_scheduler_index].get_last_lr()
```
Comment (imichalakTT) [inline tt-train/sources/ttml/ttml/common/schedulers.py]: Is `return self._last_lr` not enough?

**Review by imichalakTT — APPROVED** [review]

**Review by jmalone-tt — APPROVED** [review]

**Review by dyurshevichTT — APPROVED** [review]
